### ISSUE TITLE
Validate Rancher runtime resources supported on k8s 1.22 clusters

### ISSUE THREAD

**Issue by snasovich** [rancher/rancher#33222]
Rancher creates resources during runtime (e.g. CRDs). Update any referenced API versions that are no longer supported in k8s 1.22 (see https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22) to ones that are supported.

Following the changes, vanilla Rancher should start on k8s 1.21 with no deprecation warnings about resources to be removed in 1.22. Currently, there are many messages `apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition` logged.

Notes:
* Usages of `apiextensions.k8s.io/v1beta1` need to be changed to `apiextensions.k8s.io/v1` which is supported since k8s 1.16. We also then need to specify the minimum k8s version on our Helm charts. Note that v2.6 release will support k8s 1.18+, so using something that's available since 1.16 is fine.
* Addressing Ingress is specifically out of scope for this issue and is split into its own issue: https://github.com/rancher/rancher/issues/33312

**Comment by snasovich** [comment]
CRDs created by Wrangler are already `v1` and don't trigger the warning.
However, there is still a reference to API removed in v1.22 - https://github.com/rancher/wrangler/blob/57e55381d3ed37a7842b77cca747eccb0890fe32/pkg/generated/controllers/extensions/v1beta1/ingress.go. This will need to be investigated.

Meanwhile, it looks like v1beta1 CRDs are coming from Norman, looking into it.

**Comment by snasovich** [comment]
Norman does indeed create CRDs via deprecated API: https://github.com/rancher/norman/blob/59b3523c3133434a3943f05cbd8cc6409ea976d5/store/crd/init.go#L13

So far, locally changed the code to use `v1` API instead of `v1beta1`, however the most notable change in `v1` API is that `OpenAPIV3Schema` is now required and there doesn't seem to be code that outputs schema in this format in Norman. Such (or very close) code exists in Wrangler, will look into it.

**Task 2:** Schema for CRDs in Norman.

Meanwhile, specified open schema like this:
```go
					Schema: &apiext.CustomResourceValidation {
						OpenAPIV3Schema: &apiext.JSONSchemaProps{
							Type: "object",
							Properties: map[string]apiext.JSONSchemaProps{
								"spec": {
									XPreserveUnknownFields: &[]bool{true}[0],
								},
								"status": {
									XPreserveUnknownFields: &[]bool{true}[0],
								},
							},
						},
					},
```

Most warnings of deprecated CRD APIs went away, however, 2 are still there. Investigating where they are coming from.

**Task 3:** Investigate remaining CRD deprecation warnings.

**Comment by snasovich** [comment]
Update: These 2 deprecation warnings seems to be coming from Lasso: https://github.com/rancher/lasso/blob/fc3ebd901c086dc8e4bfe2c07edf03c118572924/pkg/dynamic/gvks.go#L37

**Comment by snasovich** [comment]
Fix to Lasso should be simply changing v1beta1 to v1 on https://github.com/rancher/lasso/blob/fc3ebd901c086dc8e4bfe2c07edf03c118572924/pkg/dynamic/gvks.go#L37.

As for schema for Norman CRDs, after discussing with further with @dramich and @ibuildthecloud using "catch-all" schema is acceptable option for now. Eventually, we would like to review these Norman CRDs and remove / migrate to Wrangler / define schema as necessary.

Continuing investigating use of removed APIs for Ingress. One item to consider here is that the replacement API for Ingress is available from k8s 1.19 only but we still need to support k8s 1.18 for v2.6 release.

**Comment by snasovich** [comment]
Also, there is usage of removed APIs in tests (**Task 4**).
https://github.com/rancher/rancher/blob/6d516b6208a7752203dba2871afdaff123d97cfa/pkg/controllers/management/test/e2e/authz_test.go
https://github.com/rancher/rancher/blob/6d516b6208a7752203dba2871afdaff123d97cfa/pkg/controllers/management/test/e2e/common_test.go

**Comment by snasovich** [comment]
After making some changes in Lasso/Norman/Wrangler, discovered that Rancher UI shows the following on landing page: _"It looks like this is your first time visting the Rancher UI, but the local admin account password is already set to something unique. Log in with that account below to continue the setup process."_ This error seems to be caused by the changes as it no longer shows if updated Lasso/Norman/Wrangler are used (after resetting cluster and deleting management-state).

**Comment by snasovich** [comment]
Resolved the issue referenced in the previous comment.

Basically, for v1 CRDs we have to define schema and in my changes to Norman I’ve used “catch-all” schema based on the following code from Wrangler:
```go
	// add a dummy schema because v1 requires OpenAPIV3Schema to be set
	if crd.Spec.Versions[0].Schema == nil {
		crd.Spec.Versions[0].Schema = &apiext.CustomResourceValidation{
			OpenAPIV3Schema: &apiext.JSONSchemaProps{
				Type: "object",
				Properties: map[string]apiext.JSONSchemaProps{
					"spec": {
						XPreserveUnknownFields: &[]bool{true}[0],
					},
					"status": {
						XPreserveUnknownFields: &[]bool{true}[0],
					},
				},
			},
		}
	}
```
However, some (or all?) CRDs (e.g. `authconfigs.management.cattle.io`) do not use `spec` and `status` and put properties directly onto root. So this schema was resulting in these fields being dropped.

So I’ve relaxed it to be even more of catch-all and it’s working now.
```go
					Schema: &apiext.CustomResourceValidation {
						OpenAPIV3Schema: &apiext.JSONSchemaProps{
							Type: "object",
							XPreserveUnknownFields: &[]bool{true}[0],
						},
					},
```

**Comment by snasovich** [comment]
Mentioned tests have been migrated to v1. Also, v1beta1 from Wrangler doesn't seem to be used, so it can probably be removed.

So far all removed k8s 1.22 APIs have been addressed (locally, work in progress) except for Ingress. There are a lot of usages of v1beta1 Ingress within Rancher and since replacement v1 API is only available from k8s v1.19, we cannot simply change them as Rancher v2.6 should still support k8s v1.18. To be discussed/decided what should be done about it.

**Comment by snasovich** [comment]
### Root cause
Rancher is using `v1beta1` version of CustomResourceDefinition object which is removed in k8s 1.22.
 
### What was fixed, or what changes have occurred
Switched to `v1` version of CustomResourceDefinition which is available since k8s 1.16.
 
### Areas or cases that should be tested
Creation of CRDs and querying of objects defined by these CRDs.
 
### What areas could experience regressions ?
All - this is a rather core change. I believe it's best to not test this issue individually, but indirectly as part of overall k8s 1.22 support effort.
 
### Are the repro steps accurate/minimal ?
No repro steps, not a bug.

**Comment by sowmyav27** [comment]
Validated deploying clusters, v2 apps on clusters with 1.22.3rancher1-1. In general things are working on 1.22.3-rancher1-1. Any bugs found have been logged separately. Closing this issue
